contracts/contracts/token/OUSD.sol

```diff
@@ -511,93 +491,185 @@ contract OUSD is Initializable, InitializableERC20Detailed, Governable {
+    }
 
-        // Increase non rebasing supply
-        nonRebasingSupply = nonRebasingSupply.add(balanceOf(msg.sender));
-        // Set fixed credits per token
-        nonRebasingCreditsPerToken[msg.sender] = _rebasingCreditsPerToken;
+    function _rebaseOptOut(address _account) internal {
+        require(
+            alternativeCreditsPerToken[_account] == 0,
+            "Account must be rebasing"
+        );
+        RebaseOptions state = rebaseState[_account];
+        require(
+            state == RebaseOptions.StdRebasing || state == RebaseOptions.NotSet,
+            "Only standard rebasing accounts can opt out"
+        );
+
+        uint256 oldCredits = _creditBalances[_account];
+        uint256 balance = balanceOf(_account);
+
+        // Account
+        rebaseState[_account] = RebaseOptions.StdNonRebasing;
+        alternativeCreditsPerToken[_account] = 1e18;
+        _creditBalances[_account] = balance;
 
-        // Decrease rebasing credits, total supply remains unchanged so no
-        // adjustment necessary
-        _rebasingCredits = _rebasingCredits.sub(_creditBalances[msg.sender]);
+        // Globals
+        nonRebasingSupply += balance;
+        _rebasingCredits -= oldCredits;
 
-        // Mark explicitly opted out of rebasing
-        rebaseState[msg.sender] = RebaseOptions.OptOut;
-        emit AccountRebasingDisabled(msg.sender);
+        emit AccountRebasingDisabled(_account);
     }
 
     /**
      * @dev Modify the supply without minting new tokens. This uses a change in
      *      the exchange rate between "credits" and OUSD tokens to change balances.
      * @param _newTotalSupply New total supply of OUSD.
      */
-    function changeSupply(uint256 _newTotalSupply)
-        external
-        onlyVault
-        nonReentrant
-    {
-        require(_totalSupply > 0, "Cannot increase 0 supply");
+    function changeSupply(uint256 _newTotalSupply) external onlyVault {
+        require(totalSupply > 0, "Cannot increase 0 supply");
 
-        if (_totalSupply == _newTotalSupply) {
+        if (totalSupply == _newTotalSupply) {
             emit TotalSupplyUpdatedHighres(
-                _totalSupply,
+                totalSupply,
                 _rebasingCredits,
                 _rebasingCreditsPerToken
             );
             return;
         }
 
-        _totalSupply = _newTotalSupply > MAX_SUPPLY
+        totalSupply = _newTotalSupply > MAX_SUPPLY
             ? MAX_SUPPLY
             : _newTotalSupply;
 
-        _rebasingCreditsPerToken = _rebasingCredits.divPrecisely(
-            _totalSupply.sub(nonRebasingSupply)
-        );
+        _rebasingCreditsPerToken =
+            (_rebasingCredits * 1e18) /
+            (totalSupply - nonRebasingSupply);
 
         require(_rebasingCreditsPerToken > 0, "Invalid change in supply");
 
-        _totalSupply = _rebasingCredits
-            .divPrecisely(_rebasingCreditsPerToken)
-            .add(nonRebasingSupply);
+        totalSupply =
```

Comment:
@DanielVF do you think we need to re-calculate the totalSupply again here? To make up for that 1 potential wei rounding error?

Reply:
Sadly, there's no way to have a perfect total supply that matches all balances. One problem is that the credits of several small accounts will convert to a different sum(balanceOf()) when rounded individually, vs when combined first, then rounded.

My actual lean here would be to keep the totalSupply unmodified (removing line 548), and accept a gap between the sum of accounts and the totalSupply, which then gets used on the next rebase. I'm not sure if there would be other ramifications though.

Reply:
just to confirm did you mean removing this bit: 
```
totalSupply =
            ((_rebasingCredits * 1e18) / _rebasingCreditsPerToken) +
            nonRebasingSupply;
```

that re-calculated the totalSupply just before the event `TotalSupplyUpdatedHighres` gets emitted? (that was my suggestion as well)

Reply:
Code simplification FTW: https://github.com/OriginProtocol/origin-dollar/pull/2298/commits/3d03c7b2287f23fd43dad6003670283a39b95e0f